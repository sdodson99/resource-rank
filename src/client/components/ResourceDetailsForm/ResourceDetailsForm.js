import React from 'react';
import ErrorAlert from '../Alerts/ErrorAlert/ErrorAlert';
import TextInput from '../TextInput/TextInput';
import PropTypes from 'prop-types';
import styles from './ResourceDetailsForm.module.css';
import LoadingSpinner from '../LoadingSpinner/LoadingSpinner';
import hasAlphaNumericCharacter from '@/validators/alpha-numeric';
import isProfane from '@/validators/profanity';
import { isURL } from 'validator';
import { useFormContext } from 'react-hook-form';
import Link from '../Link/Link';

const ResourceDetailsForm = ({
  onSubmit,
  onInvalid,
  cancelHref,
  errorMessage,
  nameFieldName,
  linkFieldName,
  isSubmitting,
}) => {
  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useFormContext();

  const nameFieldOptions = {
    required: 'Required',
    maxLength: {
      value: 50,
      message: 'Must be less than 50 characters',
    },
    validate: {
      hasAlphaNumericCharacter: (name) =>
        hasAlphaNumericCharacter(name) ||
        'Must contain an alpha-numeric character',
      isNotProfane: (name) => !isProfane(name) || 'Must not contain profanity',
    },
  };

  const linkFieldOptions = {
    required: 'Required',
    validate: {
      isURL: (link) => isURL(link) || 'Must be a valid URL',
    },
  };

  const nameError = errors[nameFieldName]?.message;
  const linkError = errors[linkFieldName]?.message;

  return (
    <div
      className={styles.ResourceDetailsForm}
      data-testid="ResourceDetailsForm"
    >
      <form onSubmit={handleSubmit(onSubmit, onInvalid)}>
        {errorMessage && (
          <div className="mb-10">
            <ErrorAlert border={true} scrollTo={!!errorMessage}>
              {errorMessage}
            </ErrorAlert>
          </div>
        )}

        <div>
          <TextInput
            name={nameFieldName}
            label="Name"
            errorMessage={nameError}
            autoComplete="off"
            {...register(nameFieldName, nameFieldOptions)}
          />
        </div>

        <div className="mt-6">
          <TextInput
            name={linkFieldName}
            label="Link"
            errorMessage={linkError}
            autoComplete="off"
            {...register(linkFieldName, linkFieldOptions)}
          />
        </div>

        <div className="mt-10 flex flex-col sm:flex-row">
          <button
            className="btn btn-primary w-100"
            type="submit"
            disabled={isSubmitting}
          >
            Submit
          </button>
          <Link url={{ pathname: cancelHref }}>
            <a className="mt-3 text-center btn btn-cancel-outline w-100 sm:mt-0 sm:ml-3">
              Cancel
            </a>
          </Link>

          {isSubmitting && (
            <div className="mt-5 sm:mt-0 sm:ml-3 self-center">
              <LoadingSpinner size={30} />
            </div>
          )}
        </div>
      </form>
    </div>
  );
};

ResourceDetailsForm.propTypes = {
  onSubmit: PropTypes.func,
  onInvalid: PropTypes.func,
  cancelHref: PropTypes.string,
  errorMessage: PropTypes.string,
  nameFieldName: PropTypes.string,
  linkFieldName: PropTypes.string,
  isSubmitting: PropTypes.bool,
};

ResourceDetailsForm.defaultProps = {};

export default ResourceDetailsForm;
