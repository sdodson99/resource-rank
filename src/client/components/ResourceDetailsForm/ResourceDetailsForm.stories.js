import React from 'react';
import { FormProvider, useForm } from 'react-hook-form';
import ResourceDetailsForm from './ResourceDetailsForm';

export default {
  component: ResourceDetailsForm,
  title: 'Components/ResourceDetailsForm', 
  // Fixes a storybook issue for maximum call stack exceeding.
  parameters: { docs: { source: { type: 'code' } } },
  decorators: [
    (Story) => {
      const methods = useForm({
        mode: 'onBlur',
        reValidateMode: 'onChange',
      });

      return (
        <FormProvider {...methods}>
          <Story />
        </FormProvider>
      );
    },
  ],
};

const Template = (args) => <ResourceDetailsForm {...args} />;

export const Default = Template.bind({});
Default.args = {
  nameFieldName: 'name',
  linkFieldName: 'link',
  cancelHref: '/',
};

export const Error = Template.bind({});
Error.args = {
  ...Default.args,
  errorMessage: 'Failed to submit form.',
};
