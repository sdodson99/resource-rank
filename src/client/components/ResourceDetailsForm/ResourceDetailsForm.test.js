import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { createRenderer } from 'react-test-renderer/shallow';
import ResourceDetailsForm from './ResourceDetailsForm';
import { useFormContext } from 'react-hook-form';
import hasAlphaNumericCharacter from '@/validators/alpha-numeric';
import isProfane from '@/validators/profanity';
import { isURL } from 'validator';

jest.mock('react-hook-form');
jest.mock('@/validators/alpha-numeric');
jest.mock('@/validators/profanity');
jest.mock('validator');

describe('<ResourceDetailsForm />', () => {
  let props;
  let formContext;

  beforeEach(() => {
    props = {
      onSubmit: jest.fn(),
      onInvalid: jest.fn(),
      cancelHref: '/back',
      errorMessage: '',
      nameFieldName: 'NAME',
      linkFieldName: 'LINK',
    };

    formContext = {
      handleSubmit: jest.fn(),
      register: jest.fn(),
      formState: {
        errors: {},
      },
    };
    useFormContext.mockReturnValue(formContext);
  });

  afterEach(() => {
    useFormContext.mockReset();
    hasAlphaNumericCharacter.mockReset();
    isProfane.mockReset();
    isURL.mockReset();
  });

  it('should mount', () => {
    render(<ResourceDetailsForm {...props} />);

    const component = screen.getByTestId('ResourceDetailsForm');

    expect(component).toBeInTheDocument();
  });

  describe('name field', () => {
    let actualNameFieldName;
    let actualNameFieldOptions;

    beforeEach(() => {
      render(<ResourceDetailsForm {...props} />);

      actualNameFieldName = formContext.register.mock.calls[0][0];
      actualNameFieldOptions = formContext.register.mock.calls[0][1];
    });

    it('should register', () => {
      expect(actualNameFieldName).toBe(props.nameFieldName);
    });

    it('should register with required validation', () => {
      expect(actualNameFieldOptions.required).toBeDefined();
    });

    it('should register with max length validation', () => {
      expect(actualNameFieldOptions.maxLength.value).toBe(50);
    });

    it('should register with alpha-numeric validation', () => {
      hasAlphaNumericCharacter.mockReturnValue(false);

      const message =
        actualNameFieldOptions.validate.hasAlphaNumericCharacter('value');

      expect(message).toBeDefined();
    });

    it('should register with profanity validation', () => {
      isProfane.mockReturnValue(true);

      const message = actualNameFieldOptions.validate.isNotProfane('value');

      expect(message).toBeDefined();
    });
  });

  describe('link field', () => {
    let actualLinkFieldName;
    let actualLinkFieldOptions;

    beforeEach(() => {
      render(<ResourceDetailsForm {...props} />);

      actualLinkFieldName = formContext.register.mock.calls[1][0];
      actualLinkFieldOptions = formContext.register.mock.calls[1][1];
    });

    it('should register', () => {
      expect(actualLinkFieldName).toBe(props.linkFieldName);
    });

    it('should register with required validation', () => {
      expect(actualLinkFieldOptions.required).toBeDefined();
    });

    it('should register with URL validation', () => {
      isURL.mockReturnValue(false);

      const message = actualLinkFieldOptions.validate.isURL('value');

      expect(message).toBeDefined();
    });
  });

  it('should render correctly', () => {
    const component = createRenderer().render(
      <ResourceDetailsForm {...props} />
    );

    expect(component).toMatchSnapshot();
  });

  it('should render correctly with error message', () => {
    props.errorMessage = 'Error message.';

    const component = createRenderer().render(
      <ResourceDetailsForm {...props} />
    );

    expect(component).toMatchSnapshot();
  });

  it('should render correctly when submitting', () => {
    props.isSubmitting = true;

    const component = createRenderer().render(
      <ResourceDetailsForm {...props} />
    );

    expect(component).toMatchSnapshot();
  });
});
