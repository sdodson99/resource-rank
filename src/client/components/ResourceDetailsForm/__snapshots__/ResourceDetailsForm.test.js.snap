// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ResourceDetailsForm /> should render correctly 1`] = `
<div
  className="ResourceDetailsForm"
  data-testid="ResourceDetailsForm"
>
  <form>
    
    <div>
      <ForwardRef(WrappedTextInput)
        autoComplete="off"
        label="Name"
        name="NAME"
      />
    </div>
    <div
      className="mt-6"
    >
      <ForwardRef(WrappedTextInput)
        autoComplete="off"
        label="Link"
        name="LINK"
      />
    </div>
    <div
      className="mt-10 flex flex-col sm:flex-row"
    >
      <button
        className="btn btn-primary w-100"
        disabled={false}
        type="submit"
      >
        Submit
      </button>
      <Link
        href="/back"
      >
        <a
          className="mt-3 text-center btn btn-cancel-outline w-100 sm:mt-0 sm:ml-3"
        >
          Cancel
        </a>
      </Link>
    </div>
  </form>
</div>
`;

exports[`<ResourceDetailsForm /> should render correctly when submitting 1`] = `
<div
  className="ResourceDetailsForm"
  data-testid="ResourceDetailsForm"
>
  <form>
    
    <div>
      <ForwardRef(WrappedTextInput)
        autoComplete="off"
        label="Name"
        name="NAME"
      />
    </div>
    <div
      className="mt-6"
    >
      <ForwardRef(WrappedTextInput)
        autoComplete="off"
        label="Link"
        name="LINK"
      />
    </div>
    <div
      className="mt-10 flex flex-col sm:flex-row"
    >
      <button
        className="btn btn-primary w-100"
        disabled={true}
        type="submit"
      >
        Submit
      </button>
      <Link
        href="/back"
      >
        <a
          className="mt-3 text-center btn btn-cancel-outline w-100 sm:mt-0 sm:ml-3"
        >
          Cancel
        </a>
      </Link>
      <div
        className="mt-5 sm:mt-0 sm:ml-3 self-center"
      >
        <LoadingSpinner
          color="#292929"
          size={30}
        />
      </div>
    </div>
  </form>
</div>
`;

exports[`<ResourceDetailsForm /> should render correctly with error message 1`] = `
<div
  className="ResourceDetailsForm"
  data-testid="ResourceDetailsForm"
>
  <form>
    <div
      className="mb-10"
    >
      <ErrorAlert
        border={true}
        scrollTo={true}
      >
        Error message.
      </ErrorAlert>
    </div>
    <div>
      <ForwardRef(WrappedTextInput)
        autoComplete="off"
        label="Name"
        name="NAME"
      />
    </div>
    <div
      className="mt-6"
    >
      <ForwardRef(WrappedTextInput)
        autoComplete="off"
        label="Link"
        name="LINK"
      />
    </div>
    <div
      className="mt-10 flex flex-col sm:flex-row"
    >
      <button
        className="btn btn-primary w-100"
        disabled={false}
        type="submit"
      >
        Submit
      </button>
      <Link
        href="/back"
      >
        <a
          className="mt-3 text-center btn btn-cancel-outline w-100 sm:mt-0 sm:ml-3"
        >
          Cancel
        </a>
      </Link>
    </div>
  </form>
</div>
`;
